File: 3630-total-characters-in-string-after-transformations-ii/total-characters-in-string-after-transformations-ii.java
class Solution {
    static final int MOD = (int) 1e9 + 7;

    public int lengthAfterTransformations(String s, int t, List<Integer> nums) {
        long[] freq = new long[26];
        for (char c : s.toCharArray()) {
            freq[c - 'a']++;
        }

        long[][] trans = new long[26][26];
        for (int i = 0; i < 26; i++) {
            int steps = nums.get(i);
            for (int j = 1; j <= steps; j++) {
                int to = (i + j) % 26;
                trans[i][to] = (trans[i][to] + 1) % MOD;
            }
        }

        long[][] transT = matrixPower(trans, t);

        long[] resultFreq = new long[26];
        for (int i = 0; i < 26; i++) {
            for (int j = 0; j < 26; j++) {
                resultFreq[j] = (resultFreq[j] + freq[i] * transT[i][j]) % MOD;
            }
        }

        long total = 0;
        for (int i = 0; i < 26; i++) {
            total = (total + resultFreq[i]) % MOD;
        }

        return (int) total;
    }

    private long[][] multiply(long[][] A, long[][] B) {
        long[][] C = new long[26][26];
        for (int i = 0; i < 26; i++) {
            for (int k = 0; k < 26; k++) {
                if (A[i][k] == 0) continue;
                for (int j = 0; j < 26; j++) {
                    C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD;
                }
            }
        }

        return C;
    }

    private long[][] matrixPower(long[][] base, int power) {
        long[][] result = new long[26][26];

        for (int i = 0; i < 26; i++) {
            result[i][i] = 1;
        }

        while (power > 0) {
            if ((power & 1) == 1) {
                result = multiply(result, base);
            }
            base = multiply(base, base);
            power >>= 1;
        }
        
        return result;
    }
}

File: 3630-total-characters-in-string-after-transformations-ii/total-characters-in-string-after-transformations-ii.go
const MOD int64 = 1e9 + 7

func lengthAfterTransformations(s string, t int, nums []int) int {
	// Bước 1: vector tần suất ký tự
	freq := make([]int64, 26);
	for _, c := range s {
		freq[c-'a']++;
	}

	// Bước 2: xây ma trận chuyển tiếp 26x26
	trans := makeMatrix();
	for i := 0; i < 26; i++ {
		steps := nums[i];
		for j := 1; j <= steps; j++ {
			to := (i + j) % 26;
			trans[i][to] = (trans[i][to] + 1) % MOD;
		}
	}

	// Bước 3: lũy thừa ma trận trans^t
	transT := matrixPower(trans, t);

	// Bước 4: nhân vector freq với ma trận kết quả
	resultFreq := make([]int64, 26);
	for i := 0; i < 26; i++ {
		for j := 0; j < 26; j++ {
			resultFreq[j] = (resultFreq[j] + freq[i]*transT[i][j]) % MOD;
		}
	}

	// Bước 5: tổng kết quả
	var total int64 = 0;
	for i := 0; i < 26; i++ {
		total = (total + resultFreq[i]) % MOD;
	}

	return int(total);
}

// Tạo ma trận 26x26 rỗng
func makeMatrix() [][]int64 {
	mat := make([][]int64, 26);
	for i := range mat {
		mat[i] = make([]int64, 26);
	}

	return mat;
}

// Nhân 2 ma trận 26x26
func multiply(A, B [][]int64) [][]int64 {
	C := makeMatrix();
	for i := 0; i < 26; i++ {
		for k := 0; k < 26; k++ {
			if A[i][k] == 0 {
				continue;
			}

			for j := 0; j < 26; j++ {
				C[i][j] = (C[i][j] + A[i][k]*B[k][j]) % MOD;
			}
		}
	}

	return C;
}

// Lũy thừa ma trận A^power
func matrixPower(base [][]int64, power int) [][]int64 {
	res := makeMatrix();
	// Ma trận đơn vị
	for i := 0; i < 26; i++ {
		res[i][i] = 1;
	}

	for power > 0 {
		if power%2 == 1 {
			res = multiply(res, base);
		}

		base = multiply(base, base);
		power /= 2;
	}

	return res;
}
